aws/aws.go

```diff
@@ -49,6 +61,104 @@ func AllASGInstanceIds(as []AutoScalingGroup) map[reapable.Region]map[reapable.I
 	return inASG
 }
 
+func AllCloudformations() chan *Cloudformation {
+	ch := make(chan *Cloudformation)
+	// waitgroup for all regions
+	wg := sync.WaitGroup{}
+	for _, region := range config.Regions {
+		wg.Add(1)
+		go func(region string) {
+			// add region to waitgroup
+			api := cloudformation.New(&aws.Config{Region: region})
+			err := api.DescribeStacksPages(&cloudformation.DescribeStacksInput{}, func(resp *cloudformation.DescribeStacksOutput, lastPage bool) bool {
+				for _, stack := range resp.Stacks {
+					ch <- NewCloudformation(region, stack)
+				}
+				// if we are at the last page, we should not continue
+				// the return value of this func is "shouldContinue"
+				if lastPage {
+					// on the last page, finish this region
+					wg.Done()
+					return false
+				}
+				return true
+			})
+			if err != nil {
+				// probably should do something here...
+				log.Error(err.Error())
+				// don't wait if the API call failed
+				wg.Done()
+			}
+		}(region)
+	}
+	go func() {
+		// in a separate goroutine, wait for all regions to finish
+		// when they finish, close the chan
+		wg.Wait()
+		close(ch)
+
+	}()
+	return ch
+}
+
+func CloudformationResources(c Cloudformation) chan *cloudformation.StackResource {
+	ch := make(chan *cloudformation.StackResource)
+	api := cloudformation.New(&aws.Config{Region: string(c.Region)})
+	// TODO: stupid
+	stringName := string(c.ID)
+
+	go func() {
+		<-timeout
+
+		// this query can fail, so we retry
+		didRetry := false
+		input := &cloudformation.DescribeStackResourcesInput{StackName: &stringName}
+
+		// initial query
+		resp, err := api.DescribeStackResources(input)
+		for err != nil {
+			sleepTime := 2*time.Second + time.Duration(rand.Intn(2000))*time.Millisecond
+			if err != nil {
+				// this error is annoying and will come up all the time... so you can disable it
+				if strings.Split(err.Error(), ":")[0] == "Throttling" && log.Extras() {
+					log.Warning(fmt.Sprintf("StackResources: %s (retrying %s after %ds)", err.Error(), c.ID, sleepTime*1.0/time.Second))
+				} else if strings.Split(err.Error(), ":")[0] != "Throttling" {
+					// any other errors
+					log.Error(fmt.Sprintf("StackResources: %s (retrying %s after %ds)", err.Error(), c.ID, sleepTime*1.0/time.Second))
+				}
+			}
+
+			// wait a random amount of time... hopefully long enough to beat rate limiting
+			time.Sleep(sleepTime)
+
+			// retry query
+			resp, err = api.DescribeStackResources(input)
+			didRetry = true
+		}
+		if didRetry && log.Extras() {
+			log.Notice("Retry succeeded for %s!", c.ID)
+		}
+		for _, resource := range resp.StackResources {
+			ch <- resource
+		}
+		close(ch)
+	}()
+	return ch
+}
+
+func ASGInstanceIDs(a *AutoScalingGroup) map[reapable.Region]map[reapable.ID]bool {
```

Comment:
Added a function that takes a single ASG instead of an array of ASGs

TODO: remove the old one that takes an array